File: usaco/section_3.3/shopping.cpp
/*
ID: huangch7
LANG: C++
TASK: shopping
*/

#include <iostream>
#include <fstream>
#include <unordered_map>
#include <cmath>
#include <algorithm>

using namespace std;

const int MAX_S = 1000;
const int MAX_B = 5;
const int MAX_C = 6;
// const int INF = 1e9;

int special_offers[MAX_S][MAX_C];
int prices[MAX_B];
int needs[MAX_B];
unordered_map<int, int> id_map;
unordered_map<int, int> res;

int get_id(int c) {
    if (id_map.find(c) == id_map.end()) {
        id_map[c] = id_map.size();
    }
    return id_map[c];
}

int merge(int alist[MAX_B]) {
    int res = 0;
    for (int i = 0; i < MAX_B; ++i) {
        res *= 10;
        res += alist[i];
    }
    return res;
}

void split(int x, int res[MAX_B]) {
  fill(res, res + MAX_B, 0);
    int p = MAX_B - 1;
    while (x) {
        res[p] = x % 10;
        x /= 10;
        p--;
    }
}

int dp(int target_list[MAX_B], int S) {
    int target_num = merge(target_list);
    if (target_num == 0) {
        return 0;
    }
    if (res.find(target_num) != res.end()) {
        return res[target_num];
    }
    int best = 0;
    for (int i = 0; i < MAX_B; ++i) {
        best += target_list[i] * prices[i];
    }
    for (int i = 0; i < S; ++i) {
        int left[MAX_B] = {0};
        bool valid = true;
        for (int j = 0; j < MAX_B; ++j) {
            left[j] = target_list[j] - special_offers[i][j];
            if (left[j] < 0) {
                valid = false;
                break;
            }
        }
        if (!valid) {
            continue;
        }
        best = min(best, dp(left, S) + special_offers[i][MAX_C - 1]);
    }
    res[target_num] = best;
    return best;
}

int main() {
    ifstream fin("shopping.in");
    ofstream fout("shopping.out");

    int S, B;
    fin >> S;
    for (int i = 0; i < S; ++i) {
        int n, p;
        fin >> n;
        for (int j = 0; j < n; ++j) {
            int c, k;
            fin >> c >> k;
            special_offers[i][get_id(c)] = k;
        }
        fin >> p;
        special_offers[i][MAX_C - 1] = p;
    }

    fill(prices, prices + MAX_B, 0);
    fill(needs, needs + MAX_B, 0);

    fin >> B;
    for (int i = 0; i < B; ++i) {
        int c, k, p;
        fin >> c >> k >> p;
        prices[get_id(c)] = p;
        needs[get_id(c)] = k;
    }
    // for (int i = 0; i < MAX_B; ++i){
    //   cout<< prices[i] << " ";
    // }
    // cout <<endl;
    // for (int i = 0; i < MAX_B; ++i){
    //   cout<< needs[i] << " ";
    // }
    // cout <<endl;
    // for (int i = 0; i < S; ++i){
    //   for (int j = 0; j < MAX_C; ++j){
    //     cout<< special_offers[i][j] << " ";
    //   }
    // cout <<endl;      
    // }


    fout << dp(needs, S) << "\n";

    fin.close();
    fout.close();

    return 0;
}

File: usaco/training/section_3.3/shopping.py
"""
ID: huangch7
LANG: PYTHON3
TASK: shopping
"""


task_name = "shopping"

def get_ints_from_line():
    return [int(x) for x in fin.readline().strip().split()]


fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

S = int(fin.readline())

special_offers = []
prices = [0] * 5
needs = [0] * 5
id_map = {}

def get_id(c):
    if c not in id_map:
        id_map[c] = len(id_map)
    return id_map[c]

for _ in range(S):
    cols = get_ints_from_line()
    n = cols[0]
    p = cols[-1]
    packages = [0] * 5
    
    for i in range(1, n + 1):
#         print(cols[i * 2 - 1], get_id(cols[i * 2 - 1]), cols[i * 2])
        packages[get_id(cols[i * 2 - 1])] = cols[i * 2]
    special_offers.append((packages, p))

B = int(fin.readline())
for _ in range(B):
    c, k, p = get_ints_from_line()
    prices[get_id(c)] = p
    needs[get_id(c)] = k

# print(special_offers)
# print(prices, needs)

# dp[a][b][c][d][e] = dp[a - pa][...] + p

def merge(alist):
    res = 0
    for x in alist:
        res *= 10
        res += x
    return res

def split(x):
    res = [0] * 5
    p = len(res) - 1
    while x:
        res[p] = x % 10
        x = x // 10
        p -= 1
    return res

res = {0: 0}

def dp(target_list):
    target_num = merge(target_list)
    if target_num in res:
        return res[target_num]
    best = 0
    for i in range(5):
        best += target_list[i] * prices[i]
    for packages, price in special_offers:
        left = [0] * 5
        valid = True
        for i in range(5):
            left[i] = target_list[i] - packages[i]
            if left[i] < 0:
                valid = False
        if not valid:
            continue
        best = min(best, dp(left) + price)
    res[target_num] = best
    return res[target_num]

fout.write(f"{dp(needs)}\n")
    
fout.close()
